### PR TITLE
fix: Dispose codec after completing frame creation

### PR DESCRIPTION
**Pull request by koji-1009**
ref https://github.com/flutter/flutter/issues/93404, [comment](https://github.com/flutter/flutter/issues/93404#issuecomment-2525453113) and [comment](https://github.com/flutter/flutter/issues/93404#issuecomment-2547622795).

Added a process to call `ui.Codec`'s `dispose`.
With this change, [HtmlBlobCodec](https://github.com/flutter/flutter/blob/8e0993eda8b0bbf0e11162cf18df1effd134db09/engine/src/flutter/lib/web_ui/lib/src/engine/html_image_element_codec.dart#L100)'s dispose will be called in Safari when using CanvasKit, and `revokeObjectURL` will be executed as expected.

* https://api.flutter.dev/flutter/dart-ui/Codec/dispose.html
* https://bugs.webkit.org/show_bug.cgi?id=31253
* https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL_static
* https://developer.mozilla.org/en-US/docs/Web/API/URL/revokeObjectURL_static

If this fix looks good, I will open the https://github.com/flutter/flutter/pull/161481 to remove the `AlearmClock` from the `BrowserImageDecoder`.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by koji-1009** [comment]
As far as I read the code, there are no rules for `ui.Codec`'s dispose process.
There is a process to dispose the codec when the process fails as follows. However, there is no process to dispose the codec when the process succeeds.

https://github.com/flutter/engine/blob/3.27.1/lib/web_ui/lib/painting.dart#L702

The `BrowserImageDecoder` caches the decoder and does not expect to be called dipose from outside as long as the process succeeds. On the other hand, `HtmlImageElementCodec` expects that dispose will be called when processing is complete to prevent leaks.

https://github.com/flutter/engine/blob/3.27.1/lib/web_ui/lib/src/engine/image_decoder.dart#L29
https://github.com/flutter/engine/blob/3.27.1/lib/web_ui/lib/src/engine/html_image_element_codec.dart#L15

If dispose is called externally when the process fails, shouldn't dispose also be called when the process succeeds?
There are no rules regarding dispose in `ui.Codec`, so even if the dispose method is properly implemented, there is still a leak.

**Comment by goderbauer** [comment]
FYI @polina-c maybe the [Codec](https://main-api.flutter.dev/flutter/dart-ui/Codec-class.html) object is one that could benefit from leak tracking? See also the issues linked from the PR description.

File: packages/flutter/lib/src/painting/image_stream.dart [hunk 1]
```diff
@@ -1057,7 +1057,11 @@ class MultiFrameImageStreamCompleter extends ImageStreamCompleter {
       final int completedCycles = _framesEmitted ~/ _codec!.frameCount;
       if (_codec!.repetitionCount == -1 || completedCycles <= _codec!.repetitionCount) {
         _decodeNextFrameAndSchedule();
+        return;
       }
+
+      _codec!.dispose();
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/image_stream.dart:1063]: We should document on the `MultiFrameImageStreamCompleter` that it takes over ownership of the Codec that was passed in and will dispose it if it is done with it.

File: packages/flutter/lib/src/painting/image_stream.dart [hunk 2]
```diff
@@ -1091,6 +1095,11 @@ class MultiFrameImageStreamCompleter extends ImageStreamCompleter {
       );
       return;
     }
+    if (_codec == null) {
+      // codec was disposed during getNextFrame
+      return;
+    }
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/image_stream.dart:1101]: The life cycle of the codec in this class is kinda unclear and it is hard to follow. For example, how can `getNextFrame` cause the disposal of the codec? Can we simplify the life cycle to make it easier to reason about to avoid future bugs?

**Comment by goderbauer** [comment]
This PR will also have to be rebased to tip of tree to adhere to the new formatting style.

**Comment by goderbauer** [comment]
@koji-1009 Have you looked at the failing golden tests (see bot comment above)? Those seem unexpected...

**Comment by koji-1009** [comment]
Repsitry was in the process of transitioning to mono-repo, so work was suspended.
I did not expect golden test to fail. I will push the fixes in order and check where golden test starts to fail.

File: packages/flutter/lib/src/painting/image_stream.dart (hunk 2 above)
Comment (koji-1009) [reply packages/flutter/lib/src/painting/image_stream.dart:1115]: This behavior is tested in the “Decoding does not crash when disposed” test case. If a listener is removed during decoding and the listeners are empty, `_maybeDispose()` is executed and the codec is disposed.

https://github.com/flutter/flutter/blob/314870064a3a45e245be749cacb1375e643d0d91/packages/flutter/test/painting/image_stream_test.dart#L149

The codec lifecycle is confusing. Now that the codec is a mutable property, it needs to be checked after asynchronous work.

**Comment by polina-c** [comment]
> FYI @polina-c maybe the [Codec](https://main-api.flutter.dev/flutter/dart-ui/Codec-class.html) object is one that could benefit from leak tracking? See also the issues linked from the PR description.

@goderbauer, yes, it looks it will.

Will make it happen as soon as monorepo lands.

**Comment by koji-1009** [comment]
I am thinking of separating PR to add `codec.dispose()` to benchmarks, golden tests, etc. In particular, since the golden test was an unexpected fail, I intend to investigate while reducing the scope of the change.

Also, the process of [BrowserImageDecoder](https://github.com/flutter/flutter/blob/8e0993eda8b0bbf0e11162cf18df1effd134db09/engine/src/flutter/lib/web_ui/lib/src/engine/image_decoder.dart#L29) caching the decoder would be unnecessary if `codec.dispose()` is called correctly. I would like to leave this discussion outside the scope of this PR, as https://github.com/flutter/flutter/issues/127548 seems appropriate for this discussion.

---

Please let me know if I should prioritize fixing the benchmark or golden test before merging this PR.

What I would like to do with this PR is to resolve the `ui.Codec` dispose issue before the HTML renderer is removed. wasm is not available in Safari. Therefore, I would like to avoid a situation where the HTML renderer is removed before this problem is resolved.

**Comment by chinmaygarde** [comment]
Triage: @polina-c @goderbauer I believe there are open questions for you. Can you please provide guidance when able please? Thanks.

**Comment by koji-1009** [comment]
If I need to make an issue, please let me know.

**Comment by polina-c** [comment]
> If I need to make an issue, please let me know.

Reviewed the PRs. Thank you for putting them together and sorry for delay.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Review by goderbauer — APPROVED** [review]
LGTM
